airflow/providers/amazon/aws/operators/lambda_function.py

```diff
@@ -62,13 +61,15 @@ class LambdaCreateFunctionOperator(BaseOperator):
     :param aws_conn_id: The AWS connection ID to use
     """
 
+    aws_hook_class = LambdaHook
     template_fields: Sequence[str] = (
         "function_name",
         "runtime",
         "role",
         "handler",
         "code",
         "config",
+        *AwsBaseOperator.template_fields,
```

Comment:
Rather than adding this to every operator class, can we do it the other way around where the base class has this field already with these defaults and we update it with anything new that the concrete class wants to add (if anything)? It would need to be a mutable type of course, I'm not sure if it being immutable is a requirement or not.

Reply:
I think we could update it only in init method, at least I don't know another way to change class attribute during creation, and I'm not sure is it works on each cases or not (especially for mapped tasks).


Another way it is make parameters as set, but this have side effect, set doesn't preserve the order, and it might generate new serialised DAG each time it parsed (maybe it is not so bad)

```python
class Base:
    templates_fields: set[str] = {"a", "b", "c"}

class Child1(Base):
    ...

class Child2(Base):
    templates_fields = Base.templates_fields | {"d", "e"}

class Child3(Base):
    ...

print(" Step 1 ".center(72, "="))
print(f"Base.templates_fields={Base.templates_fields}")
print(f"Child1.templates_fields={Child1.templates_fields}")
print(f"Child2.templates_fields={Child2.templates_fields} - the order would change from run to run")
print(f"Child3.templates_fields={Child3.templates_fields}")

print(" Step 2: Update one of the child ".center(72, "="))
Child3.templates_fields.update({"f", "g"})

print(f"Base.templates_fields={Base.templates_fields}")
print(f"Child1.templates_fields={Child1.templates_fields}")
print(f"Child2.templates_fields={Child2.templates_fields} - the order would change from run to run")
print(f"Child3.templates_fields={Child3.templates_fields}")

print(" Step 3: Invalid operation ".center(72, "="))
class Child5(Base):
    templates_fields.add("h")  # We can't do that
```

```console
================================ Step 1 ================================
Base.templates_fields={'b', 'c', 'a'}
Child1.templates_fields={'b', 'c', 'a'}
Child2.templates_fields={'e', 'b', 'd', 'c', 'a'} - the order would change from run to run
Child3.templates_fields={'b', 'c', 'a'}
=================== Step 2: Update one of the child ====================
Base.templates_fields={'b', 'f', 'g', 'c', 'a'}
Child1.templates_fields={'b', 'f', 'g', 'c', 'a'}
Child2.templates_fields={'e', 'b', 'd', 'c', 'a'} - the order would change from run to run
Child3.templates_fields={'b', 'f', 'g', 'c', 'a'}
====================== Step 3: Invalid operation =======================
Traceback (most recent call last):
...
NameError: name 'templates_fields' is not defined
```

Reply:
I just thought about another option, create helper function for apply default parameters

```python
from airflow.compat.functools import cache

@cache
def aws_template_fields(*templated_fields: str) -> tuple[str]:
    return tuple(sorted(list({"aws_conn_id", "region_name", "verify"} | set(templated_fields))))


class SomeOperator:
    template_fields: Sequence[str] = aws_template_fields(
        "function_name",
        "runtime",
        "role",
        "handler",
        "code",
        "config",
   )
```

WDYT?

Reply:
A variation on that might be to have each operator define "op_template_fields" and have BaseOperator's `template_fields` return base+op?  Basically:

```
Class BaseOperator:
    base_template_fields = ("aws_conn_id", "region_name", "verify")

    @cache
    def template_fields:
        return self.base_template_fields + self.operator_template_fields
        
        
Class MyOperator:
    op_template_fields =  ("foo", ''"bar")
```

Reply:
> I just thought about another option, create helper function for apply default parameters

This is certainly nicer! Less duplication and the defaults are complexity hidden from the users. It would be nice to have a fully backwards compatible approach, but I'm happy to settle on this one.

> A variation on that might be to have each operator define "op_template_fields"

This would make the existing operators incompatible since the field name would need to change, but we need to make other changes to them to convert them to the base class approach, so maybe that's perfectly fine?

Reply:
I've added this helper https://github.com/apache/airflow/pull/34890/commits/28aef3b617a246255a0f6a71b32371a55807b603